oculus_ros2/include/oculus_ros2/sonar_viewer.hpp

```diff
@@ -0,0 +1,138 @@
+
+#include <cv_bridge/cv_bridge.h>
+#include <oculus_driver/AsyncService.h>
+#include <oculus_driver/SonarDriver.h>
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <oculus_interfaces/msg/ping.hpp>
+#include <oculus_ros2/conversions.hpp>
+#include <opencv2/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <rclcpp/rclcpp.hpp>
+#include <sensor_msgs/msg/image.hpp>
+#include <std_msgs/msg/header.hpp>
+
+class SonarViewer {
+public:
+  // SonarViewer(bool *arg);
+  explicit SonarViewer(rclcpp::Node* node);
+  ~SonarViewer();
+  void publish_fan(const oculus::PingMessage::ConstPtr& ping, const std::string& frame_id = "", const int& data_depth = 0) const;
+  void publish_fan(
+      const oculus_interfaces::msg::Ping& ros_ping_msg, const std::string& frame_id = "", const int& data_depth = 0) const;
+  template <typename dataType>
+  void publish_fan(const int& width,
+      const int& height,
+      const int& offset,
+      const std::vector<dataType>& ping_data,
+      const int& master_mode,
+      const double& ping_rage,
+      const std_msgs::msg::Header& header) const;
+  void stream_and_filter(const oculus::PingMessage::ConstPtr& ping, cv::Mat& data);
+  // void stream_and_filter(const oculus::PingMessage::ConstPtr &ping);
+  // sensor_msgs::msg::Image publish_fan(const oculus::PingMessage::ConstPtr &ping);
+
+private:
+  //      //   sensor_msgs::ImagePtr msg;
+  const rclcpp::Node* node_;
+  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_publisher_;
+};
+
+template <typename T>
+std::vector<double> linspace(T start_in, T end_in, int num_in) {
+  std::vector<double> linspaced;
+
+  double start = static_cast<double>(start_in);
+  double end = static_cast<double>(end_in);
+  double num = static_cast<double>(num_in);
+
+  if (num == 0) {
+    return linspaced;
+  }
+  if (num == 1) {
+    linspaced.push_back(start);
+    return linspaced;
+  }
+
+  double delta = (end - start) / (num - 1);
+
+  for (int i = 0; i < num - 1; ++i) {
+    linspaced.push_back(start + delta * i);
+  }
+  linspaced.push_back(end);  // I want to ensure that start and end
+                             // are exactly the same as the input
+  return linspaced;
+}
+
+template <typename dataType>
+void SonarViewer::publish_fan(const int& width,
+    const int& height,
+    const int& offset,
+    const std::vector<dataType>& ping_data,
+    const int& master_mode,
+    const double& ping_range,
+    const std_msgs::msg::Header& header) const {
+  // Create rawDataMat from ping_data
+  cv::Mat rawDataMat(height, (width + 4), CV_8U);
+  std::memcpy(rawDataMat.data, ping_data.data() + offset, height * (width + 4));
+
+  int bearing = (master_mode == 1) ? 65 : 40;
+  std::vector<double> ranges = linspace(0., ping_range, height);
+  int image_width = 2 * std::sin(bearing * M_PI / 180) * ranges.size();
+  cv::Mat mono_img;
+  if (std::is_same<dataType, uint8_t>::value) {
```

Comment:
```cpp
if constexpr (std::is_same<dataType, uint8_t>::value)
```

is probably better, as this will be chosen at compile time and not at runtime

Reply:
Use `std::numeric_limits<uint8/16_t>::max();` to avoid magic numbers

Reply:
I hadn't see your comment, so I used `1 << sizeof(DataType) * CHAR_BIT`. I think std::numeric_limits is doing the same.